linkerd/app/outbound/src/http/logical/policy/route/backend.rs

```diff
@@ -39,6 +39,20 @@ struct BackendError {
     source: Error,
 }
 
+/// Synthesizes 504 Gateway Timeout responses for the `timeouts.backend_request`
+/// timeout.
+///
+/// This is necessary because we want these timeouts to be retried, and the
+/// retry layer only retries requests that fail with an HTTP status code, rather
+/// than for requests that fail with Rust `Err`s. Errors returned by the
```

Comment:
If we're doing this to satisfy a deficiency of the retry layer, is there any reason that we can't move this mapping closer to that decision?  I.e. why does this belong in the backend stack? Why does the backend stack care about the retry layer?

Furthermore, how does this interact with error metrics? How/where are request timeout errors recorded? Is it consistent with other error metrics?

Reply:
> If we're doing this to satisfy a deficiency of the retry layer, is there any reason that we can't move this mapping closer to that decision? I.e. why does this belong in the backend stack? Why does the backend stack care about the retry layer?

This is a good point, IIRC, I don't think there's any particular reason for this to be in the backend stack rather than in the route stack right below the retry layer. I'll see about moving it.

> Furthermore, how does this interact with error metrics? How/where are request timeout errors recorded? Is it consistent with other error metrics?

Regarding error metrics, I believe we don't currently track error metrics per-backend, only request count metrics. Therefore, the timeout will only be recorded by error metrics if it is not retried and the request fails. Because the design for HTTPRoute outbound policy metrics is still ongoing, I didn't implement any new metrics as part of this PR. When we do add HTTPRoute-specific error metrics, I would expect these timeouts to be recorded by any per-backend error counters but not any per-route error counters if the timeout is retried.